domaci_bandit.py

```diff
@@ -98,54 +99,113 @@ def train(bandits_no=5, attempts_no=5000, alpha=0.1, epsilon=0.1, plotting=True)
 # Prilikom smanjenja vrednosti epsilona, cesce je birana masina sa najvecom srednjom vrednosti (samim tim i veca nagarda) te je nagib
 # krive kumulativne nagrade rastao.
 
 
 # Zadatak 2 : Sa naucenim q i epsilon = 0, pustiti 100 iteracija
-def drugi_zadatak():
-    plt.figure(figsize=(8, 3))
-
-    envi, rew, q1 = train(plotting=False)
-    plot_e(2, 1, envi, rew)
-    plt.title("5000 iteracija")
-
-    plt.subplot(1, 2, 2)
-    test_len = 100
-    reward_q1 = [max(q1) for _ in range(test_len)]  # Uvek ce biti ista nagrada
+def drugi_zadatak(plotting=False):
+    env, rewards, q = train(plotting=False)
+    learned_q = q.copy()
+
+    # 100 iteracija sa epsilon = 0 (greedy metoda)
+    greedy_rewards = []
+    for _ in range(100):
+        a = choose_greedy_action(learned_q)
+        r = env.take_action(a)
+        greedy_rewards.append(r)
 
-    g1 = np.concatenate((rew, reward_q1))
-    plot_e(2, 2, envi, g1)
-    plt.title("5100 iteracija")
-    plt.show()
+    avg_rewards_eps_greedy = np.mean(rewards)
+    avg_rewards_greedy = np.mean(greedy_rewards)
+    print(f"Max Q naučena vrednost: {max(learned_q):.2f}")
+    print("Prosečna nagrada - ε-Greedy Politika:", avg_rewards_eps_greedy)
+    print("Prosečna nagrada - Čista Greedy Politika:", avg_rewards_greedy)
+    if plotting:
+        last_100_eps_greedy_rewards = rewards[-100:]
+
+        cumulative_rewards_eps_greedy = np.cumsum(last_100_eps_greedy_rewards)
+        cumulative_rewards_greedy = np.cumsum(greedy_rewards)
+
+        plt.figure(figsize=(12, 6))
+        plt.plot(
+            cumulative_rewards_eps_greedy,
+            label="Kumulativne Nagrade - Poslednjih 100 ε-Greedy",
+        )
+        plt.plot(
+            cumulative_rewards_greedy,
+            label="Kumulativne Nagrade - Greedy",
+            linestyle="--",
+        )
+
+        plt.xlabel("Epoha")
+        plt.ylabel("Kumulativna Nagrada")
+        plt.title("Poredjenje Kumulativnih Nagrada: ε-Greedy vs Greedy")
+        plt.legend()
+        plt.show()
+
+        # Poredjenje na poslednjih 10 nagrada
+        max_mean_reward = max(bandit.mean for bandit in env.bandits)
+        eps_greedy_every_10 = last_100_eps_greedy_rewards[::10]
+        greedy_every_10 = greedy_rewards[::10]
+
+        epochs = range(len(rewards) - 100, len(rewards), 10)
+        plt.scatter(
+            epochs,
+            eps_greedy_every_10,
+            label="ε-Greedy Nagrade (svaka 10ta)",
+            marker="o",
+            alpha=0.7,
+        )
+
+        plt.scatter(
+            epochs,
+            greedy_every_10,
+            label="Greedy Nagrade (svaka 10ta)",
+            marker="x",
+            alpha=0.7,
+        )
+
+        plt.axhline(
+            y=max_mean_reward,
+            color="r",
+            linestyle="--",
+            label="Max Mean Moguća Nagrada",
+        )
+
+        plt.xlabel("Epoha")
+        plt.ylabel("Nagrada")

```

Comment:
ovaj ceo drugi grafik mi je jusles iskreno... ne vidim nista iz njega :)

fali garfik koji sam ja napisala, msm da je to i najbitniji bio

Reply:
Kako je jusles? Taj je jedini i bitan, da vidis kako odlucuje epsilon a kako gridi politika u svakom momentu. Bas dobro vizualizuje prednosti jednog i drugog